mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,549 @@
+         */
+        case id: cp.Ident if id.symbol == NoSymbol =>
+          fallbackSymbol(id.name, id.pos) match {
+            case Some(sym) => nodes + new NodeInfo(id, id.pos)
+            case _ => nodes
+          }
+
+        case df: cp.MemberDef =>
+          (tree.children ++ annotationChildren(df))
+            .foldLeft(nodes)(traverse(_, _))
+        case _ =>
+          if (tree==null)null
+          else tree.children.foldLeft(nodes)(traverse(_, _))
+      }
+    }
+
+    def fallbackSymbol(name: Name, pos: Position) = {
+      val context = cp.doLocateImportContext(pos)
+      context.lookupSymbol(name, sym => sym.isType) match {
+        case LookupSucceeded(_, symbol) =>
+          Some(symbol)
+        case _ => None
+      }
+    }
+    private def annotationChildren(mdef: cp.MemberDef): List[cp.Tree] = {
+      mdef.mods.annotations match {
+        case Nil if mdef.symbol != null =>
+          // After typechecking, annotations are moved from the modifiers
+          // to the annotation on the symbol of the annotatee.
+          mdef.symbol.annotations.map(_.original)
+        case anns => anns
+      }
+    }
+
+    private def typePos(tpe: cp.TypeTree) = {
+      tpe.original match {
+        case cp.AppliedTypeTree(tpt, _) => tpt.pos
+        case sel: cp.Select => sel.namePos
+        case _ => tpe.pos
+      }
+    }
+    // We need to collect named params since they will not show on fully typed tree
+    lazy val namedArgCache = {
+      val parsedTree = cp.parseTree(source)
+      parsedTree.collect { case arg @ cp.AssignOrNamedArg(_, rhs) =>
+        rhs.pos.start -> arg
+      }.toMap
+    }
+  }
+
+  def pickFromTraversed(tk:scala.meta.tokens.Token):List[cp.Tree] ={
+      val buffer = ListBuffer.empty[cp.Tree]
+        
+      for (node <- nodes){
+        if( node.pos.start ==tk.pos.start &&
+           node.pos.end ==tk.pos.end 
+        ) buffer.addAll(List(node.tree))
+      }
+
+      buffer.toList
+  }
+
+
+  /**
+    * returns (SemanticTokenType, SemanticTokenModifier) of @param tk
+    */
+  private def getTypeAndMod(tk:scala.meta.tokens.Token):(Int, Int,String) ={
+
+    // whether token is identifier or not
+    tk match {
+      case _: Token.Ident => // continue this method
+      case _ =>
+         //Non-Ident has no modifier.
+         return (typeOfNonIdentToken(tk), 0, strSep + "Non-Ident")
+    }
+
+    var logString = ""
+    logString += linSep + linSep  + "  Start:Ident Part getSemanticTypeAndMod"
+    logString += linSep + "  " + tk.name
+
+    val nodeList = pickFromTraversed(tk)
+    val node = if (nodeList.size ==0) null else nodeList(0)
+    if( node == null) return (-1,0,strSep + "Node-Nothing") // break
+
+    logString += linSep + "  ** Got node"
+    //get type
+    val sym = node.symbol
+    val typ =  if (sym.isValueParameter ) getTid(SemanticTokenTypes.Parameter)
+      else node.symbol.keyString match {
+          case kind.kDef => 
```

Comment:
Could we just just if it's a method? `symbol.isMethod` ?

I would not match on string if possible

Reply:
I will do that.

Reply:
Sorry, but I still want to match with _Symbol.KeyString_. 
If I match with flags, the result will be sensitive to the order of the if statement. It seems danger.
_Symbol.keyString_ absorbs that. For example, in the following image, "val" contains the condition that it is not a "var" (!isVariable).
![image](https://user-images.githubusercontent.com/92659783/190704875-2fd0bbf8-3b00-40a0-8242-41e6ec168d53.png)